packages/transformers/src/lit-transformer.ts

```diff
@@ -0,0 +1,263 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import * as ts from 'typescript';
+import {LitElementMutations} from './mutations.js';
+
+type Visitor = ClassDecoratorTransformer | MemberDecoratorTransformer;
+
+interface ClassDecoratorTransformer {
+  kind: 'classDecorator';
+  decoratorName: string;
+
+  visit(
+    mutations: LitElementMutations,
+    class_: ts.ClassDeclaration,
+    decorator: ts.Decorator
+  ): void;
+}
+
+interface MemberDecoratorTransformer {
+  kind: 'memberDecorator';
+  decoratorName: string;
+
+  visit(
+    mutations: LitElementMutations,
+    member: ts.ClassElement,
+    decorator: ts.Decorator
+  ): void;
+}
+
+const unreachable = (x: never) => x;
+
+/**
+ * Configurable transformer for LitElement classes.
+ */
+export class LitTransformer {
+  private _context: ts.TransformationContext;
+
+  private _classDecoratorVisitors = new Map<
+    string,
+    ClassDecoratorTransformer[]
+  >();
+
+  private _memberDecoratorVisitors = new Map<
+    string,
+    MemberDecoratorTransformer[]
+  >();
+
+  private _removeNodes = new Set<ts.Node>();
+
+  constructor(context: ts.TransformationContext, visitors: Array<Visitor>) {
+    this._context = context;
+    for (const visitor of visitors) {
+      switch (visitor.kind) {
+        case 'classDecorator': {
+          let arr = this._classDecoratorVisitors.get(visitor.decoratorName);
+          if (arr === undefined) {
+            arr = [];
+            this._classDecoratorVisitors.set(visitor.decoratorName, arr);
+          }
+          arr.push(visitor);
+          break;
+        }
+
+        case 'memberDecorator': {
+          let arr = this._memberDecoratorVisitors.get(visitor.decoratorName);
+          if (arr === undefined) {
+            arr = [];
+            this._memberDecoratorVisitors.set(visitor.decoratorName, arr);
+          }
+          arr.push(visitor);
+          break;
+        }
+
+        default:
+          throw new Error(
+            `Internal error: unknown visitor kind ${
+              (unreachable(visitor) as Visitor).kind
+            }`
+          );
+      }
+    }
+  }
+
+  visit = (node: ts.Node): ts.VisitResult<ts.Node> => {
+    if (this._removeNodes.delete(node)) {
+      return undefined;
+    }
+    if (ts.isImportDeclaration(node)) {
+      return this._visitImportDeclaration(node);
+    }
+    if (ts.isClassDeclaration(node)) {
+      return this._visitClassDeclaration(node);
+    }
+    return ts.visitEachChild(node, this.visit, this._context);
+  };
+
+  private _visitImportDeclaration(node: ts.ImportDeclaration) {
+    if (
+      ts.isStringLiteral(node.moduleSpecifier) &&
+      node.moduleSpecifier.text === 'lit/decorators.js'
```

Comment:
Consider also checking for `'lit-element'` since that's probably still common, and `'`lit/decorators'` since that's how auto-imports will tend to get written

Reply:
Done. Actually I think all of these are valid (!) so I'm mostly using prefix tests now:

```
'lit/decorators.js',
'lit/decorators',
'lit/decorators/custom-element.js',
'lit/decorators/custom-element',
'@lit/reactive-element/decorators.js',
'@lit/reactive-element/decorators',
'@lit/reactive-element/decorators/custom-element.js',
'@lit/reactive-element/decorators/custom-element',
'lit-element',
'lit-element/index.js',
'lit-element/index',
'lit-element/decorators.js',
'lit-element/decorators',
```

Reply:
The non-file-extension imports are not valid actually. They could be used by toolchains that don't (yet) support package exports, but I don't think we should support that. Those projects will break and have to fix imports when they update tools.

Reply:
> The non-file-extension imports are not valid actually. They could be used by toolchains that don't (yet) support package exports, but I don't think we should support that. Those projects will break and have to fix imports when they update tools.

If we don't support them, the behavior will be that we silently don't transform decorators if they are imported without an extension. Is that your preference? We could also have an explicit check for that style of import, but here seems like the wrong place to do that.

Reply:
I feel like we should error here. At first we'll be the ones using this and our code should be correct. If we find that we see valid requests to support extension-less imports, and it somehow would be better to fail, and it guaranteed to fail, somewhere else in the pipeline, then we could relax the strictness here. I don't want to perpetuate tools that do the wrong thing and will allow code that breaks when tools are upgraded.

Reply:
Done. We now error if you don't import with  `.js` extension or the default module.